File: Data Structure/1918번 - 후위 표기식/1918.cpp
#include <iostream>
#include <stack>

using namespace std;

string infix = "";

void input() {
    cin >> infix;
}

int pri(char ch) {
    switch(ch) {
        case '(': case ')': return 0;
        case '+': case '-': return 1;
        case '*': case '/': return 2;
    }
    return -1;
}

void solve() {
    stack<char> S;
    char st_top;
    
    for(int i = 0; i < infix.length(); i++) {
        char cur_w = infix[i];
        if(cur_w == '(')
            S.push(cur_w);
        else if(cur_w == ')') {
            st_top = S.top(); S.pop();
            while(st_top != '(') {
                cout << st_top;
                st_top = S.top(); S.pop();
            }
        }
        else if(cur_w == '+' || cur_w == '-' || cur_w == '*' || cur_w == '/') {
            while(!S.empty() && (pri(cur_w) <= pri(S.top()))) {
                st_top = S.top(); S.pop();
                cout << st_top;
            }
            S.push(cur_w);
        }
        else {
            cout << cur_w;
        }
    }
    
    while(!S.empty()) {
        cout << S.top(); S.pop();
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Data Structure/1918번 - 후위 표기식/1918.py
import sys

infix = input().rstrip()

def pri(ch):
    if ch == '(' or ch == ')':
        return 0
    elif ch == '+' or ch == '-':
        return 1
    elif ch == '*' or ch == '/':
        return 2

def solve():
    ans = ''
    stack = []
    
    for c in infix:
        if c == '(':
            stack.append('(')
        elif c == ')':
            t = stack.pop()
            while stack and t != '(':
                ans += t
                t = stack.pop()
        elif c == '+' or c =='-' or c == '*' or c == '/':
            while stack and pri(c) <= pri(stack[-1]):
                ans += stack.pop()
            stack.append(c)
        else:
            ans += c
            
    stack.reverse()
    print(ans + ''.join(stack))


if __name__ == '__main__':
    solve()